presto-main/src/main/java/com/facebook/presto/operator/MultiChannelGroupByHash.java

```diff
@@ -204,40 +205,33 @@ public void appendValuesTo(int groupId, PageBuilder pageBuilder, int outputChann
     public void addPage(Page page)
     {
         if (canProcessDictionary(page)) {
-            addDictionaryPage(page);
+            boolean done = new AddDictionaryWork(page).process();
+            // TODO: (1) change the interface and (2) remove this with yield enabled
+            verify(done);
             return;
         }
 
-        // get the group id for each position
-        int positionCount = page.getPositionCount();
-        for (int position = 0; position < positionCount; position++) {
-            // get the group for the current row
-            putIfAbsent(position, page);
-        }
+        boolean done = new AddPageWork(page).process();
```

Comment:
* AddPageWork/AddDictionaryWork
* GetDictionaryGroupIdsWork/GetPageGroupIdsWork

I don't think this is a good way to distinct dictionary vs not dictionary. It's not dictionary vs page. I would rather a longer name to avoid this trickiness.